asynq/scheduler.py

```diff
@@ -60,262 +55,94 @@ def __init__(self):
+            self._continue_with_batch()
 
-    def add_dependencies(self, task, yield_result):
-        """Mark all futures in yield_result as dependencies for task.
+    def _execute(self, root_task):
+        """Implements task execution loop.
 
-        i.e. everything in yield_result needs to be computed before task can be continued.
+        The postcondition of this method is that all tasks in the dependency tree of root_task
+        that aren't blocked on batch items waiting to be flushed should be executed until they are
+        (or until they're computed). This is done by running a depth-first search on the dependency
+        tree.
 
+        :param root_task: root of the dependency tree
+        :return: ``None``
         """
-        if yield_result is None:  # Frequent case
-            return
-        for future in extract_futures(yield_result, []):
-            self.add_dependency(task, future)
-
-    def add_dependency(self, task, dependency):
-        if dependency.is_computed():
-            return
-        if dependency in task._dependencies:
-            if _debug_options.DUMP_DEPENDENCIES:
-                debug.write('@async: +dependency: %s already depends on %s' %
-                    (debug.str(task), debug.str(dependency)))
-            return
-        if _debug_options.DUMP_DEPENDENCIES:
-            debug.write('@async: +dependency: %s needs %s' %
-                (debug.str(task), debug.str(dependency)))
-        dependency.make_dependency(task, self)
-
-    def _await(self, tasks):
-        active_task = self.active_task
-        if active_task is not None:
-            # Since task we await for can be (and most likely is)
-            # a child of the current one, it already copied the
-            # list of its contexts; so when we start it, these
-            # contexts are going to be resumed, but we don't want
-            # to resume already active contexts. That's why we
-            # need to pause them.
-            active_task._pause_contexts()
-            # We must temporarily block active task in case
-            # it's somehow (c) added to scheduler's queue
-            blocking_dependency = _create_blocking_dependency()
-            active_task._dependencies.add(blocking_dependency)
-        try:
-            # Must never throw an exception
-            if len(tasks) == 0:
-                # No arguments = run() must empty task queue
-                return self._execute(False, self._empty_set)
+        init_num_tasks = len(self._tasks)
+        self._tasks.append(root_task)
+
+        # Run the execution loop until the root_task is complete (it's either blocked on batch
+        # items waiting to be flushed, or computed).
+        while len(self._tasks) > init_num_tasks:
+            if len(self._tasks) > MAX_TASK_RECURSION_DEPTH:
+                self.reset()
+                debug.dump(self)
+                assert False, 'maximum recursion depth exceeded while calling an asynq task'
```

Comment:
`raise RuntimeError` for consistency with what happens if you overflow the non-asynq stack